### PR TITLE
Refactor project and add basic shell completion

### PR DESCRIPTION
**Pull request by cfi2017**
I've gone through the points I made on discord a couple days ago and implemented the most glaring changes:
- you don't (read shouldn't) have a go.mod file in every one of your subdirectories  
=> I've reduced the entire structure down to a single go.mod file in the root directory to simplify dependency management.
- I'd rename all your main.go files to something more descriptive (eg config.go for configuration) as main usually denotes an entrypoint  
=> I've removed one redundant (empty) main.go file and refactored the others to match their respective package name.
- I'd move your structures to a model package or something similar in either the internal package or, better, the pkg package. That allows people to easily find models (atm the mod struct is defined in mod/main.go, the registry struct in game/mod_registry.go, moving it makes it clearer)  
=> I have not yet looked at this, due to complexity of the refactor and the models being a little more complex than I initially thought.
- I'd keep my tests close to the actual code files they test. You can keep helpers in the test files, they won't be compiled with the application. _test.go denotes a test file, this is treated differently by the compiler (so for example download might have a download_test file in the same directory). the test directory on the root folder is by definition more for loosely coupled test related data like third party test tools, test data and such, not necessarily for unit tests  
=> Tests are still the same, I have also not looked at test content yet
- I'd consider moving everything in your ./cmd folder into a ./internal/commands folder and then moving your root main.go file to ./cmd/gpm/main.go. This allows you to then run go build ./cmd/gpm to build the application and clearly denotes the entrypoint while also adding support for multiple entrypoints.  
=> Done, `go build ./cmd/gpm` should now result in the wanted binary.
- For releases, I find goreleaser pairs really well with github actions (see an example of it here: https://github.com/cfi2017/bl3-save). It allows you to build for a os/arch matrix so you can do multi-platform builds and automatically propagates changelogs for you (important files in the above repo are .goreleaser.yml and the github actions configuration)  
=> Goreleaser is now added, I have not however done a tagged commit to test this part of the CI.

### CODE REVIEW

File: .github/workflows/analysis.yaml [hunk 1]
```diff
@@ -1,23 +1,33 @@
-name: Analysis
+name: static check
 on:
   push:
+    tags:
+      - "v*"
```
Comment (osulli) [inline .github/workflows/analysis.yaml]: Can you change this to just wildcard please?

I prefer versioning as "x.x.x" not "vx.x.x" as it is redundant when discussing versions.

File: .github/workflows/release.yaml [hunk 2]
```diff
@@ -1,30 +1,26 @@
-name: Release
+name: goreleaser
+
 on:
   push:
-    branches:
-      - main
+    tags:
+      - 'v*.*.*'
```
Comment (osulli) [inline .github/workflows/release.yaml]: Same again here please with the v

We'll version both the release and tag itself with semver

File: .github/workflows/release.yaml [hunk 3]
```diff
@@ -1,30 +1,26 @@
...
+      - name: Unshallow
+        run: git fetch --prune --unshallow
+      - name: Set up Go
+        uses: actions/setup-go@v1
         with:
-          go-version: '1.15.4'
-      - name: Go Version
-        run: go version
-      - uses: wangyoucao577/go-release-action@v1.11
+          go-version: 1.14.x
```
Comment (osulli) [inline .github/workflows/release.yaml]: Will this be ok with the hardcoded `.x`

File: .goreleaser.yml [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+builds:
+    - env:
+          - CGO_ENABLED=0
+      goos:
+          - darwin
+          - linux
+          - windows
+      goarch:
+          - amd64
+          - 386
```
Comment (osulli) [inline .goreleaser.yml:18]: What is 386?

**Review by osulli — CHANGES_REQUESTED** [review]
That's great. Thanks for the refactor and gorelease + linter recommendations.

As discussed, please can you add the linter to the pre-commit config file + update Makefile where appropriate.

Thanks

**Review by osulli — CHANGES_REQUESTED** [review]
Also, is there a reason `go test ...` or `make test` is no longer running on PRs?

File: .goreleaser.yml (hunk 4 above)
Comment (cfi2017) [reply .goreleaser.yml:18]: 32 bit compilation, just adds to the matrix. adds a couple more build strategies

File: .github/workflows/release.yaml (hunk 3 above)
Comment (cfi2017) [reply .github/workflows/release.yaml]: Changing this to v1.15, though 1.14.x should have been fine

**Comment by cfi2017** [comment]
> Also, is there a reason `go test ...` or `make test` is no longer running on PRs?

I left that out for now until we have a more stabilised test structure. I'd like to re-add that as a step after the linting process but I haven't gotten to it yet

**Comment by MythicManiac** [comment]
You might want to direct this PR to https://github.com/WolvenKit/gpm-go now that the repos are separated

**Comment by osulli** [comment]
https://github.com/WolvenKit/gpm-go/pull/1 merged

**Pull request closed without merging**